Clarify confusing aspects of exemplar specification. (#2201)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

## Exemplar

> Exemplars are example data points for aggregated data. They provide specific

Reviewer: Q: are they always full-fidelity data points, i.e. representing all dimensions of the original measurement?

Reviewer 2: The java implementation of exemplar only includes attributes that were filtered out by the aggregator. So the combination of the metric attributes plus the exemplar attributes provides the full set originally collected. 

I believe this behavior is specified a bit [further down](https://github.com/open-telemetry/opentelemetry-specification/pull/2201/files#diff-32d5f289beed96900fa1e6e69a3ee9cc0b6503ce8883672011c44bc772c0fa2aR513-R514).

Author: The simple answer is `yes`.   The longer answer (for OTLP) is we only store 'missing' dimensions from the aggregated metric data point we're associated with.

Reviewer 3: You're suggesting a more detailed answer -- here's what I suggest:

Whether or not all dimensions are included in the exemplar, the exemplar fundamentally represent a raw, individual measurement whereas both Sum and Histogram (i.e., all synchronous) OTLP data points have effectively no way to record raw measurements.

context to otherwise general aggregations. Exemplars allow correlation between
aggregated metric data and the original API calls where measurements are
recorded. Exemplars work for trace-metric correlation across any metric, not

> just those that can also be derived from `Span`s.

Reviewer 2: > across any metric, not just those that can also be derived from `Span`s.

What exactly do you mean by this?

Author: @yurishkuro has a system where you can automatically attach Span-derived metric exemplar points.   Specifically, you could derive latency metrics from spans and if you had a way to correlate to a specific metric, you could generate exemplars directly from the span itself.   The OTEL Exemplar does this, but also in scenarios where your metric does NOT correlated with a Span attribtue (e.g. "memory allocations", if you can record in context)

> 

Reviewer: I think it's useful to provide a link to the OTEP for further background / decisions.

Author: We don't do this in any of the other parts of the specification.   Is that a change we're making in expecations?

Reviewer 3: It's very old by now, but there is one. 
https://github.com/open-telemetry/oteps/blob/main/text/metrics/0113-exemplars.md

An [Exemplar](./datamodel.md#exemplars) is a recorded
[Measurement](./api.md#measurement) that exposes the following pieces of
information:

> - The `value` that was recorded at API call time.
> - The `time` the `Measurement` was sent to the API.

Reviewer 2: "call time" and "sent to the API" dont appear to be consistent with other verbiage. A possible alternative: 
```suggestion
- The `value` of the `Measurement` that was recorded by API call.
- The `time` the API call was made to record the `Measurement`.
```

Author: I like your second option, will clarify.

- The set of [Attributes](../common/common.md#attributes) associated with the
  `Measurement` not already included in a metric data point.

>   **Note: Views may aggregate away `Attributes`, which is where this is used.**

Reviewer: Not clear - can views trim down attributes of the exemplars themselves? 

Rather than have a Note here I would move this example into the intro paragraph above.

Author: Good call out, will fix.

Author: Added an example, PTAL

just those that can also be derived from `Span`s.

An [Exemplar](./datamodel.md#exemplars) is a recorded
[Measurement](./api.md#measurement) that exposes the following pieces of
information:

- The `value` of the `Measurement` that was recorded by the API call.
- The `time` the API call was made to record a `Measurement`.
- The set of [Attributes](../common/common.md#attributes) associated with the
  `Measurement` not already included in a metric data point.
- The associated [trace id and span
  id](../trace/api.md#retrieving-the-traceid-and-spanid) of the active [Span
  within Context](../trace/api.md#determining-the-parent-span-from-a-context) of
  the `Measurement` at API call time.

For example, if a user has configured a `View` to preserve the attributes: `X`
and `Y`, but the user records a measurement as follows:

```javascript
const span = tracer.startSpan('makeRequest');
api.context.with(api.trace.setSpan(api.context.active(), span), () => {
  // Record a measurement.
  cache_miss_counter.add(1, {"X": "x-value", "Y": "y-value", "Z": "z-value"});
  ...
  span.end();
})
```

Then an examplar would consist of:

- The `value` of 1.
- The `time` when the `add` method was called

> - The `Attributes` of `{"Z": "z-value"}`, as these are not preserved in the

Reviewer: I find this example misleading. It focuses on the storage aspects of the exemplar (which can change throughout aggregation pipeline), instead of giving a logical representation of the exemplar having all 3 attributes.

Reviewer 3: Perhaps this can be clarified. "resulting metric point" below refers to one of e.g. `NumberDataPoint`, `HistogramDataPoint`, which are concrete members of the pipeline, so I think it makes sense to focus on how the protocol looks. The exemplar's attributes are nested in one of the data points, so there is no loss of information.

These are individual measurements, that is all, and the attributes (whether included or not included) are copied from actual synchronous API events. If more text is needed to explain how to merge exemplars, for example, that belongs in `datamodel.md` I think.

It's possible to add an adjusted count concept to exemplars; at that point I would call them raw samples, not exemplars. I've put a few details into https://github.com/lightstep/varopt/blob/main/README.md for how this can be done so that adjusted counts are unbiased--the result is that the samples can be used to reconstruct the aggregated metrics they are included with in a statistically meaningful way. This is a property that can be preserved through merging OTLP data points, but if we were to go this route, it would be more than exemplars.

Author: I'll clarify that this is talking the *output* in OTLP, vs. the logical representation.  This is *not* intended to represent internal storage, in fact we want to allow SDKs to store full attributes in exemplars and filter later for OTLP if that's faster.

Author: I updated the wording, PTAL.

A Metric SDK MUST provide a mechanism to sample `Exemplar`s from measurements.
- The `value` of 1.
- The `time` when the `add` method was called
- The `Attributes` of `{"Z": "z-value"}`, as these are not preserved in the
  resulting metric point.
- The trace/span id for the `makeRequest` span.

A Metric SDK MUST allow `Exemplar` sampling to be disabled.  In this instance
A Metric SDK MUST provide a mechanism to sample `Exemplar`s from measurements
via the `ExemplarFilter` and `ExemplarReservoir` hooks.

A Metric SDK MUST allow `Exemplar` sampling to be disabled. In this instance
the SDK SHOULD not have overhead related to exemplar sampling.

A Metric SDK MUST sample `Exemplar`s only from measurements within the context
of a sampled trace BY DEFAULT.

> A Metric SDK MUST allow exemplar sampling to leverage the configuration of a

Author: Not sure why it shows the line as deleted, but it's right below this one.

The "offer" method does not need to store all measurements it is given and
MAY further sample beyond the `ExemplarFilter`.

The "collect" method MUST return accumulated `Exemplar`s. Exemplars are expected
to abide by the `AggregationTemporality` of any metric point they are recorded
with. SDKs are free to decide whether "collect" should also reset internal

> storage for delta aggregation collection, or use a more optimal implementation.

Reviewer 4: It's not clear for me what it means for the exemplar to abide by the temporality of the metric point. Do you mean that CUMULATIVE shouldn't reset sampling state but DELTA should reset it?

I would think you'd want to reset regardless of temporality (even for CUMULATIVE) so that you send fresh exemplars instead of repeating the same exemplar as probability drops with the `SimpleFixedSizeExemplarReservoir`. I suppose the benefit of not resetting a CUMULATIVE is to get a true random sample over the process lifetime?

Would it be easier to have the interface provide both `collect()` and `collectAndReset()` and let the SDK determine when to reset?

Reviewer 3: > Do you mean that CUMULATIVE shouldn't reset sampling state but DELTA should reset it?

I'd like more information from Prometheus on how exemplars are being handled. I do think the `num_measurements_seen` variable should reset so that points within an interval have equal probability, but if no new exemplars arrive during  period, then the Prometheus exporter should continue to see the last exemplar.

Author: Exemplars in prometheus (did the investigation prior to writing any of this) abide by cumulative rules here.  THey keep last-seen-per-bucket always and you cannot collect exemplars for anything but histograms (even though they're attached to samples, if you use this it breaks for any metric type but histograms).

Specifically, the prometheus client library never clears the exemplars AFAICT, which means we need to decide what to do for DELTAs (as usual) but if we align with Prometheus, it means respecting the start/stop time window.

Author: Updated the verbage here, PTAL

> `Exemplar`s MUST retain the any attributes available in the measurement that

Reviewer 5: ```suggestion
`Exemplar`s MUST retain any attributes available in the measurement that
```

  end
  ```

Additionally, the `num_measurements_seen` count SHOULD be reset at every

> collection cycle.

Author: That's getting into weeds of impelmentation a bit.  From a specification standpoint this is about ensuring we get FRESH exemplars in a CUMULATIVE case.  If you're optimising your exemplar based on known CUMULATIVE reader-exporter pipeline, then you don't have to clean everything, and I'm trying to avoid forcing that in the specification.
